Create docker-proxy TCP/UDP listener sockets in the daemon

robmry opened the pull request:
**- What I did**

- Fix https://github.com/moby/moby/issues/47951
- Relates to https://github.com/moby/moby/pull/47871

Before commit 4f09af6267, when allocating host ports for a new port mapping iptables rules were set up, then docker-proxy was started. If the host port was already in-use, docker-proxy exited with an error, and the iptables rules were removed. That could potentially interfere with a non-docker service that was already using the host port for something unrelated.

Commit 4f09af6267 swapped that problem for a different one... in order to check that a port was available before creating iptables rules, it attempted to start docker-proxy first. If it failed, it could then try a different host port, without interfering with any other service. The problem with that is docker-proxy would start listening before the iptables rules were in place, so it could accept connections then become unusable because new NAT rules diverted packets directly to the container. This would leave the client with a broken connection, causing at-least a delay while it figured that out and reconnected.

**- How I did it**

This change creates and binds the socket in the daemon, before creating iptables rules. If the bind fails, it may try a different port. When or if the bind succeeds, iptables rules are created, then the daemon calls listen on the socket. If docker-proxy is needed, the socket is handed over to it at that point.

If docker-proxy is not needed (`--userland-proxy=false`), the daemon still listens on TCP sockets as the `dummyProxy` would have done. This makes the socket show up in `netstat` output.

The `dummyProxy` is no longer needed on Linux. Its job was to bind the host ports if docker-proxy was disabled, but that's now already handled by binding the sockets early.

This change doesn't affect SCTP, because it's not currently possible for docker-proxy to convert the file descriptor into an SCTPListener. So, docker-proxy is still started early, and the window for lost connections remains.

A new option has been added to `docker-proxy`, `-use-listen-fd` to tell it that fd 4 is a listener socket. It no longer needs to know the "frontend" address/port for TCP/UDP, but it's useful to see the options in the command line to work out which docker-proxy is which. (As a follow-up, it'd be possible to read the address/port from the fd and display it in the process title.)

If the user has an old docker-proxy in their path it'll fail for TCP/UDP, because it won't recognise the new command line option that says an listen-fd has been supplied as fd 4. In this case, the daemon's error message suggests checking $PATH.

RootlessKit used a proxy for `docker-proxy` called `rootlesskit-docker-proxy`. Its job was to make RootlessKit API calls to tell it about mapped host ports when necessary, before starting the usual `docker-proxy`, and then to remove that mapping when stopping the proxy. The `rootlesskit-docker-proxy` didn't know about new option `-use-listen-fd`, or the open sockets passed-in. Rather than update `rootlesskit-docker-proxy`, replace it by making the RootlessKit API calls directly from the daemon. That'll make it possible to continue to evolve `docker-proxy`, without having to keep `rootlesskit-docker-proxy` in-step (for example, we may go to a single `docker-proxy` per container rather than per port-mapping). It also makes it possible to set up those mappings in RootlessKit without running the proxy, when `--userland-proxy=false`. Because `rootlesskit-docker-proxy` is no longer used, it's no longer built.

**- How to verify it**

Updated tests.

The port mapping tests relied on `startProxy` being called to check that expected ports had been selected. That call is not made now, because`dummyProxy` isn't used. So, changed most of the tests to make them start the proxy.

For tests of behaviour where a host port is unexpectedly unavailable, a real `bind()` now needs to fail in the daemon code being tested (the test's mock proxy can't report that it would have failed, because it's called much later in the process). So, the tests now set up a dummy bridge in their n/w namespace, add any host addresses necessary to that bridge, and bind ports that need to be busy - before the test starts.

With an old `docker-proxy` in the path, the error looks like this ...

`docker: Error response from daemon: driver failed programming external connectivity on endpoint fervent_brattain (b4a9c5ebe14946f70c12c304da95abf41ed8b804903879c1163030ddbdbf1273): failed to start userland proxy for port mapping 0.0.0.0:8080:172.21.0.2:80/tcp: failed to start docker-proxy, check that the current version is in your $PATH.`

(I hacked a `NewSCTPListener(fd)` into the sctp package, and checked that it worked with the socket bound in the daemon. So we can propose a change like that, then align SCTP socket creation with TCP/UDP.)

**Rootless**

Mapping from unspecified host address ...

```
$ docker run --rm -ti --network n1 -p 8080:80 nginx
$ ps -ef | grep docker-proxy
robm       39815   39165  0 18:37 ?        00:00:00 /home/robm/bin/docker-proxy -proto tcp -host-ip 127.0.0.1 -host-port 8080 -container-ip 172.18.0.2 -container-port 80
robm       39822   39165  0 18:37 ?        00:00:00 /home/robm/bin/docker-proxy -proto tcp -host-ip ::1 -host-port 8080 -container-ip 172.18.0.2 -container-port 80

These work ...
curl http://192.168.175.43:8080 # host ip
curl http://127.0.0.1:8080
curl http://[::1]:8080
```

Mapping from a specific host address ...

```
$ docker run --rm -ti --network n1 -p 192.168.175.43:8080:80 nginx
$ ps -ef | grep docker-proxy
robm       39978   39165  0 18:39 ?        00:00:00 /home/robm/bin/docker-proxy -proto tcp -host-ip 127.0.0.1 -host-port 8080 -container-ip 172.18.0.2 -container-port 80
```

Mapping from two specific host addresses, before the change ...

```
$ docker run --rm -ti --network n1 -p 192.168.175.43:8080:80 -p 127.0.0.1:8080:80 nginx
<long delay, the 16s timeout on trying to start rootlesskit-docker-proxy>
docker: Error response from daemon: driver failed programming external connectivity on endpoint admiring_williamson (b3d757bf1d0832fcc85eccd9692c72fb24ba2ec817495e977824516cbfbdcda2): failed to bind port 192.168.175.43:8080/tcp: Timed out proxy starting the userland proxy.

<then subsequent valid requests for the same mapped port fail with an error from RootlessKit>
$ docker run --rm -ti --network n1 -p 192.168.175.43:8080:80 nginx
docker: Error response from daemon: driver failed programming external connectivity on endpoint xenodochial_taussig (24118b2865fa4fc3f314cf172c21d1118e3a28aa0cccce82d33ceff002eaf3a9): failed to bind port 192.168.175.43:8080/tcp: Error starting userland proxy: error while calling PortManager.AddPort(): conflict with ID 2.

<need to restart rootlesskit>
```

After the change - binding fails immediately, because both mappings end up trying to use the same child address ...

```
$ docker run --rm -ti --network n1 -p 192.168.175.43:8080:80 -p 127.0.0.1:8080:80 nginx
docker: Error response from daemon: driver failed programming external connectivity on endpoint charming_johnson (1d749d940f091a93108d8364ae658a9c8b44b57b362963cfc18f5e01cf904f25): failed to bind port 192.168.175.43:8080/tcp: Error starting userland proxy: failed to listen on 127.0.0.1:8080: listen tcp4 127.0.0.1:8080: bind: address already in use.
```

With `--userland-proxy=false` ...

```
$ docker run --rm -ti --network n1 -p 8080:80 nginx
-> no docker-proxy process
-> curl works for http://127.0.0.1:8080, http://192.168.175.43:8080 - but not http://[::1]:8080

$ docker run --rm -ti --network n1 -p 192.168.175.43:8080:80 nginx
-> no docker-proxy process
-> curl works for http://192.168.175.43:8080
```

**- Description for the changelog**
```markdown changelog
- Close a window in which `docker-proxy` could accept TCP connections, which would fail after NAT rules were set up. The `docker-proxy` binary has been updated, the old version will not work with the updated `dockerd`.
- The executable `rootlesskit-docker-proxy` is no longer used, it has been removed from the build.
```

robmry committed:
Fix error type used for ProgramExternalConnectivity

Get rid of "FIXME: Got an API for which error does not match
any expected type!!!  error="driver failed programming external
connectivity on endpoint..." from debug logs.

Signed-off-by: Rob Murray <rob.murray@docker.com>

corhere committed:
cmd/docker-proxy: drop unused BackendAddr() method

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
cmd/docker-proxy: drop FrontendAddr() method

It was only used in tests.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere reviewed (commented):
I have a suggestion: go all Single Responsibility Principle on the proxy implementations and make it the caller's responsibility to construct the net.Listener. https://github.com/robmry/moby/compare/bind_socket_for_docker_proxy...corhere:moby:bind_socket_for_docker_proxy?expand=1

I think docker-proxy should be backwards compatible with older dockerd. It's easy enough to do (if my suggestion is followed...) and could save users some grief if they accidentally mix versions of the binaries.

robmry committed:
cmd/docker-proxy: pass open listener to proxy impl

In preparation for the daemon passing a listen fd, add command line
option -use-listen-fd to indicate that the fd is present (as fd 4).

If the new flag isn't given, open the listener as normal.

Refactor the TCP and UDP proxies to be constructed with an existing
TCPListener or UDPConn, respectively. Lift the responsibilty of opening
the listener to the entrypoint. Per the Single Responsibility Principle,
this structure affords changing how the listener is created without
having to touch the proxy implementations.

Co-authored-by: Cory Snider <csnider@mirantis.com>
Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Create docker-proxy TCP/UDP listener sockets in the daemon

Before commit 4f09af6, when allocating host ports for a new
port mapping, iptables rules were set up then docker-proxy was
started. If the host port was already in-use, docker-proxy exited
with an error, and the iptables rules were removed. That could
potentially interfere with a non-docker service that was already
using the host port for something unrelated.

Commit 4f09af6 swapped that problem for a different one... in
order to check that a port was available before creating iptables
rules, it attempted to start docker-proxy first. If it failed, it
could then try a different host port, without interfering with
any other service. The problem with that is docker-proxy would
start listening before the iptables rules were in place, so it
could accept connections then become unusable because new NAT
rules diverted packets directly to the container. This would leave
the client with a broken connection, causing at-least a delay
while it figured that out and reconnected.

This change creates and binds the socket in the daemon, before
creating iptables rules. If the bind fails, it may try a different
port. When or if the bind succeeds, iptables rules are created,
then the daemon calls listen on the socket. If docker-proxy is
needed, the socket is handed over to it at that point.

In rootless mode, the ports have to be bound to an address in the
rootless network namespace (where dockerd is running). DNAT rules
now use the same address.

If docker-proxy is not needed ("--userland-proxy=false"), the daemon
still listens on TCP sockets as the old dummyProxy would have done.
This makes the socket show up in "netstat" output.

The dummyProxy is no longer needed on Linux. Its job was to bind the
host ports if docker-proxy was disabled, but that's now already
handled by binding the sockets early.

This change doesn't affect SCTP, because it's not currently possible
for docker-proxy to convert the file descriptor into an SCTPListener.
So, docker-proxy is still started early, and the window for lost
connections remains.

If the user has an old docker-proxy in their path and it's given a
listener docker with '-use-listen-fd', it'll fail because of the
unknown option. In this case, the daemon's error message suggests
checking $PATH.

Signed-off-by: Rob Murray <rob.murray@docker.com>

corhere commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -134,6 +146,33 @@ func (n *bridgeNetwork) addPortMappings(
+		if bindings[i].Proto == types.TCP {
+			// If not starting the proxy, nothing will ever accept a connection on the
+			// socket. But, listen anyway so that the binding shows up in "netstat -at".
+			if err := syscall.Listen(int(bindings[i].boundSocket.Fd()), somaxconn); err != nil {
```
```suggestion
			if err := syscall.Listen(int(bindings[i].boundSocket.Fd()), 0); err != nil {
```
Why tell the kernel to queue up the maximum number of connections possible, consuming kernel resources, when we have no intention to ever accept() any of them? [Zero is a valid backlog value.](https://github.com/torvalds/linux/commit/64a146513f8f)

robmry replied:
That's what `dummyProxy` did but, right, there's no need.

corhere commented on libnetwork/portmapper/proxy.go:
Rename to `proxy_windows.go`?

robmry replied:
Done.

corhere commented on libnetwork/portmapper/proxy_linux.go:
```diff
@@ -1,78 +1,64 @@
-	containerIP net.IP, containerPort int,
+// StartProxy starts the proxy process at proxyPath.
+// If listenSock is not nil, it must be a bound socket that can be passed to
+// the proxy process for it to listen on. StartProxy always takes ownership
+// of listenSock if proxyPath is nonzero, the socket will be closed even if
+// the proxy fails to start.
```
I don't think it is necessary for StartProxy to clobber listenSock. Even in the successful case where the open socket is passed into the docker-proxy subprocess, the file descriptor in the dockerd process remains a valid open reference to the underlying file description. Having StartProxy close the file descriptor is therefore _policy_, and you know what my opinions are on policy. Let the caller decide what to do with it. For example, having the caller retain ownership of the socket fd affords:
- retrying starting the proxy (after a short delay) if something goes wrong
- relaunching the proxy process with the same socket if it dies unexpectedly
- seamlessly upgrading the proxy on the fly (start new proxy process listening on socket, signal old proxy to stop accepting new connections and exit once the connection count drops to zero.)

robmry replied:
Done.

robmry commented on cmd/docker-proxy/main.go:
```diff
@@ -13,11 +13,27 @@ import (
+	// connections or receive UDP. (Without that option on the command
+	// line, the listener needs to be opened by docker-proxy, for
+	// compatibility with older docker daemons. In this case fd 4
+	// may be an epoll fd belonging to Go.)
```
Done

robmry commented:
The rootless tests are failing because, it turns out, there's a [rootlesskit-docker-proxy](https://github.com/rootless-containers/rootlesskit/blob/master/cmd/rootlesskit-docker-proxy/main.go) (a docker-proxy proxy) - and it doesn't know about the new `-use-listen-fd` option.

I think the tests passed before the last set of updates added that option, even though dockerd would already have bound the ports in the namespace it's using. Maybe dockerd needs to know if it's rootless and not-bind the ports, or just not use the new option to say it has. Or, something. I'll investigate ...

corhere commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -134,6 +146,42 @@ func (n *bridgeNetwork) addPortMappings(
+	// socket. But, listen anyway so that the binding shows up in "netstat -at".
+	somaxconn := 0
+	if proxyPath != "" {
+		somaxconn = getSOMAXCONN()
```
```suggestion
		somaxconn = -1
```
[The kernel will silently cap the backlog to SOMAXCONN for you.](https://man7.org/linux/man-pages/man2/listen.2.html#:~:text=If%20the%20backlog,the%20value%20128.) [(Kernel source)](https://elixir.bootlin.com/linux/v6.9.8/source/net/socket.c#L1876)

robmry replied:
Oh, ok - I'll change it. I wonder why `net.ListenTCP` ends up [reading the config](https://github.com/golang/go/blob/87ec2c959c73e62bfae230ef7efca11ec2a90804/src/net/sock_linux.go#L34-L54).

corhere commented:
> [rootlesskit-docker-proxy](https://github.com/rootless-containers/rootlesskit/blob/master/cmd/rootlesskit-docker-proxy/main.go) (a docker-proxy proxy)

That smells a lot like a port-forwarding "driver" for rootless Docker. I wonder if it would be worthwhile to move the rootlesskit client stuff into the dockerd side so rootlesskit-docker-proxy can go away. Dockerd is already integrating with rootlesskit in other respects, and getting rootlesskit-docker-proxy out of the way would make it that much more practical to tackle e.g. implementing proxy-per-container multiplexing.

robmry committed:
Tell RootlessKit about docker-proxy port mappings

Before this change, when running rootless, instead of running
docker-proxy the daemon would run rootlesskit-docker-proxy.

The job of rootlesskit-docker-proxy was to tell RootlessKit
about mapped host ports before starting docker-proxy, and then
to remove the mapping when it was stopped.

So, rootlesskit-docker-proxy would need to be kept in-step
with changes to docker-proxy (particuarly the upcoming change
to bind TCP/UDP ports in the daemon and pass them to the proxy,
but also possible-future changes like running proxy per-container
rather than per-port-mapping).

This change runs the docker-proxy in rootless mode, instead of
rootlesskit-docker-proxy, and the daemon itself tells RootlessKit
about changes in host port mappings.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Tell RootlessKit about port mappings when --userland-proxy=false

The daemon was modified to tell RootlessKit about host port
mappings directly, rather than by running rootlesskit-docker-proxy
to make those updates.

DNAT rules created in rootless mode referred to the host IP address,
rather than the address seen as host address in the rootless network
namespace.

With these changes, port mappings work in rootless mode when
--userland-proxy=false - so, don't gate the RootlessKit API calls
on starting docker-proxy.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Remove rootlesskit-docker-proxy

It's not needed, now the daemon tells RootlessKit about port
mappings directly.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
> > [rootlesskit-docker-proxy](https://github.com/rootless-containers/rootlesskit/blob/master/cmd/rootlesskit-docker-proxy/main.go) (a docker-proxy proxy)
> 
> That smells a lot like a port-forwarding "driver" for rootless Docker. I wonder if it would be worthwhile to move the rootlesskit client stuff into the dockerd side so rootlesskit-docker-proxy can go away. Dockerd is already integrating with rootlesskit in other respects, and getting rootlesskit-docker-proxy out of the way would make it that much more practical to tackle e.g. implementing proxy-per-container multiplexing.

Thanks Cory, I've had a go at that - there are three new commits ...
- Make the RootlessKit API calls from the daemon before starting `docker-proxy` (instead of `rootlesskit-docker-proxy`).
- Make those calls in rootless mode, even when `--userland-proxy=false`, which didn't work in rootless mode before.
- Remove `rootlesskit-docker-proxy` from the build.

AkihiroSuda commented on libnetwork/drivers/bridge/rootlesskit_client_linux.go:
```diff
@@ -0,0 +1,131 @@
+	"github.com/rootless-containers/rootlesskit/v2/pkg/port"
+)
+
+type rkPortDriverClient struct {
```
nit: s/rk/rlk/g, as it is not a rootkit...

robmry replied:
Fixed (removed `rk`, because it's now in package `rlkclient`).

AkihiroSuda commented on libnetwork/drivers/bridge/rootlesskit_client_linux.go:
```diff
@@ -0,0 +1,131 @@
+// installed as a proxy for docker-proxy:
+// https://github.com/rootless-containers/rootlesskit/blob/4fb2e2cb80bf13eb28b7f2a4317b63406b89ad32/cmd/rootlesskit-docker-proxy/main.go
+
+package bridge
```
Probably this should be a separate package

robmry replied:
Done - I made it a package under `bridge`, not sure if that was the right choice.

AkihiroSuda commented on libnetwork/portmapper/proxy_linux.go:
```diff
@@ -1,78 +1,61 @@
+		Args: []string{
+			proxyPath,
+			"-proto", pb.Proto.String(),
+			"-host-ip", pb.HostIP.String(),
```
This has to consume `childIP` chosen by the former `rootlesskit-docker-proxy` code
https://github.com/rootless-containers/rootlesskit/blob/4fb2e2cb80bf13eb28b7f2a4317b63406b89ad32/cmd/rootlesskit-docker-proxy/main.go#L200

robmry replied:
Ohhh, thank you - I'd missed that, it all makes more sense now!

(I'd only tested with no host address specified, so the proxy bound to all addresses in the child namespace. Now tested with specific host addresses as well, as shown in the description.)

`DNAT` rules now also use the host IP (which makes a port mapping with a specific host address work in rootless mode with `--userland-proxy=false`).

thaJeztah commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -447,31 +500,177 @@ func attemptBindHostPorts(
+				} else {
+					// TODO(robmry) - it's not currently possible to pass a bound SCTP port
+					//  to the userland proxy, because the proxy is not able to convert the
+					//  file descriptor into an sctp.SCTPListener (fd is an unexported member
```
Would changes to the sctp library we're using mean that we could get rid of this special case? I also noticed there's a `ListenSCTPext` variant that allows setting some options (maybe that one could be expanded?); https://github.com/ishidawataru/sctp/blob/7ff4192f6ff2b4b181ec7dce269bdf4076ff90a0/sctp_linux.go#L174

ISTR @ishidawataru is a colleague of @AkihiroSuda (but regardless, it's an open source project, so we can contribute / propose changes if that would help us)

thaJeztah commented on Dockerfile:
```diff
@@ -377,8 +377,6 @@ RUN --mount=from=rootlesskit-src,src=/usr/src/rootlesskit,rw \
   export CGO_ENABLED=$([ "$DOCKER_STATIC" = "1" ] && echo "0" || echo "1")
   xx-go build -o /build/rootlesskit -ldflags="$([ "$DOCKER_STATIC" != "1" ] && echo "-linkmode=external")" ./cmd/rootlesskit
   xx-verify $([ "$DOCKER_STATIC" = "1" ] && echo "--static") /build/rootlesskit
-  xx-go build -o /build/rootlesskit-docker-proxy -ldflags="$([ "$DOCKER_STATIC" != "1" ] && echo "-linkmode=external")" ./cmd/rootlesskit-docker-proxy
```
🎉 nice!

⚠️ looks like we need to make some changes in packaging once this is merged, as it's currently expecting this binary to be part of the packages, e.g. (probably other places as well) https://github.com/docker/docker-ce-packaging/blob/2af20a9161e87b9845ea2f6b3bfc0653c583cb25/deb/common/rules#L88

corhere commented:
> * Make those calls in rootless mode, even when `--userland-proxy=false`, **which didn't work in rootless mode before.**

(emphasis mine)

That sounds like a changelog-worthy enhancement. Yet it's nowhere to be found in the PR description. Oversight?

robmry commented:
> > * Make those calls in rootless mode, even when `--userland-proxy=false`, **which didn't work in rootless mode before.**
> 
> (emphasis mine)
> 
> That sounds like a changelog-worthy enhancement. Yet it's nowhere to be found in the PR description. Oversight?

It's mentioned in the PR description, it's an improvement that fell out naturally from these changes.

But, I haven't done any detailed testing of it in rootless mode, other than to show it's less likely to be broken if someone [just tries it](https://github.com/moby/moby/issues/43090). And, we don't properly document differences between `--userland-proxy` true/false, or limitations, even in non-rootless mode (https://github.com/docker/docs/issues/17312).

So, I don't think we're ready to announce "new thing works" in the changelog.

akerouanton reviewed (approved):
This looks pretty good! I only have a minor nit at this point, plus two things to do in follow-ups (at least if we all agree on them).

I really would like to see the 'classic port-mapping' / 'rootlesskit port-mapping' decoupling happen at some point, to allow peeling off the different layers we now have in `port_mapping_linux.go`. I think this file grew quite a lot and start to be hard to follow at times.

robmry commented on cmd/docker-proxy/network_proxy_test.go:
```diff
@@ -167,20 +182,17 @@ func testProxyAt(t *testing.T, proto string, proxy Proxy, addr string, halfClose
 	backend.Run()
 	frontendAddr := &net.TCPAddr{IP: net.IPv4(127, 0, 0, 1), Port: 0}
-	proxy, err := NewProxy(frontendAddr, backend.LocalAddr())
+	listenFd, frontendAddr := tcpListener(t, "tcp4", frontendAddr)
```
> `listenFd` isn't an 'raw' fd, it's a `*net.TCPListener`. And this makes me think that there's no tests making sure 'fd passing' (and reconstructing a listener out of it) works properly. Maybe we should have a dedicated tests 🤔

It's called `listener` now.

> On the other hand, we have `TestTCPDualStackProxy` which isn't realistic IMO. Maybe we should put everything after `parseFlags` in proxy's `main.go` into a separate func, and write tests in such a way that the proxy is started with a valid config. However, this might make this PR even more heavyweight, so maybe it's fine to do that in a follow-up.

Yes, it's worth using `newProxy` to create the `Proxy` from a `ProxyConfig`. Then, as well as checking the Go listener can be reconstructed from the socket, it can test `newProxy` creating its own TCP/UDP listener (for the case where one isn't passed in by an old version of the daemon). That's not a big change to the tests, can do it here.

Splitting `main` after the call to `parseFlags` would be more involved - it'd mean dealing with the expected pipe on `fd:3`, it starts the proxy as well as creating it, and it starts a goroutine to deal with `SIGTERM`. That stuff hasn't changed, I don't think it's worth the extra upheaval here.

robmry commented on cmd/docker-proxy/network_proxy_test.go:
```diff
@@ -192,33 +204,31 @@ func TestTCP4ProxyHalfClose(t *testing.T) {
+	testProxyAt(t, "tcp", proxy, frontendAddr.String(), false)
 }
 
 func TestTCPDualStackProxy(t *testing.T) {
```
`newProxy` can't create a dual-stack listener - and the daemon doesn't, at the moment.

But, when the socket's passed in, its creation is outside docker-proxy's control, and there's no reason it must v6-only. So, I don't think dropping the test is a good idea.

(The test was skipped before this PR, because it listened on `::1` instead of `::`. So, the socket ended up being IPv6-only and the test failed. But that's fixed now.)

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -40,21 +48,33 @@ type portBinding struct {
 	stopProxy func() error
 }
 
+// childPortBinding is pb.PortBinding, with the host address the daemon
```
Yes, that'll probably be the way we want to do it when we get to nftables, depending on the approach we take.

The interface/composition will need a bit of thought, because they're not independent (as I think you've seen in the code highlighted in this comment) - docker-proxy and iptables rules need to use a host address determined by rootlesskit.

corhere reviewed (changes requested):
I spotted some fiddly little problems with the CLOEXEC and NONBLOCK flags on the socket file descriptors which could potentially cause security issues and performance regressions, respectively.

robmry commented on libnetwork/drivers/bridge/rlkclient/rootlesskit_client_linux.go:
```diff
@@ -0,0 +1,149 @@
+		return nil, fmt.Errorf("error while connecting to RootlessKit API socket: %w", err)
+	}
+
+	info, err := c.Info(context.Background())
```
Sure - I've plumbed it back up to the bridge driver. (Where it's given a `context.TODO()`.)

robmry commented on libnetwork/drivers/bridge/rlkclient/rootlesskit_client_linux.go:
```diff
@@ -0,0 +1,149 @@
+
+	// info.PortDriver is currently nil for "none" and "implicit", but this may change in future
+	if info.PortDriver == nil || info.PortDriver.Driver == "none" || info.PortDriver.Driver == "implicit" {
+		return nil, nil
```
Done.

robmry commented on libnetwork/drivers/bridge/rlkclient/rootlesskit_client_linux.go:
Done.

robmry commented on libnetwork/drivers/bridge/rlkclient/rootlesskit_client_linux.go:
```diff
@@ -0,0 +1,149 @@
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
```
Ok.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -116,24 +153,77 @@ func (n *bridgeNetwork) addPortMappings(
+		if bindings[i].Proto == types.TCP {
+			if err := syscall.Listen(int(bindings[i].boundSocket.Fd()), somaxconn); err != nil {
+				return nil, fmt.Errorf("failed to listen on TCP socket: %w", err)
+			}
```
Done, but I'm not sure it's necessary - you're concerned about a concurrent close? Getting a blocking fd isn't an issue.

corhere replied:
Concurrent close is a small part, but I'm more interested in future-proofing against subtle changes in runtime behaviour across Go releases.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -404,30 +500,177 @@ func attemptBindHostPorts(
+		sa = &sa6
+	}
+
+	sd, err := syscall.Socket(domain, typ, proto)
```
Ok, done - but doesn't Go close it anyway? (After the exec, only `ExtraFiles` are left open in the child?)

corhere replied:
No. [Go will arrange to have the `ExtraFiles` dup'd to the expected file descriptor numbers in the child, but does not implicitly close any fd >= 3.](https://github.com/golang/go/blob/1985c0ccf9ef0736aeb79ef548689aa935fa5c4c/src/syscall/exec_linux.go#L562-L615) (There appears to be an undocumented feature of `syscall.StartProcess` in that `ProcAttr.Files[i] = -1` results in fd `i` being explicitly closed in the child, though AFAICT it is not used by os/exec.) The runtime depends on the invariant that all open file descriptors are marked close-on-exec when [`syscall.ForkLock`](https://pkg.go.dev/syscall#ForkLock) is not held for reading, and goes to [great](https://github.com/golang/go/blob/1985c0ccf9ef0736aeb79ef548689aa935fa5c4c/src/net/sock_cloexec.go) [lengths](https://github.com/golang/go/blob/1985c0ccf9ef0736aeb79ef548689aa935fa5c4c/src/net/sys_cloexec.go) to assure that for file descriptors opened by stdlib.

robmry commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -404,30 +500,177 @@ func attemptBindHostPorts(
+		domain = syscall.AF_INET6
+	}
+
+	sd, err := syscall.Socket(domain, syscall.SOCK_STREAM, syscall.IPPROTO_SCTP)
```
Ok.

robmry commented on cmd/docker-proxy/main.go:
```diff
@@ -31,41 +47,114 @@ func main() {
-	default:
-		log.Fatalf("unsupported protocol %s", *proto)
+	if useListenFd {
+		config.ListenSock = os.NewFile(listenSockFd, "listen-sock")
```
Done. I guess it'll be more significant if we make `docker-proxy` handle more than one connection.

corhere replied:
`docker-proxy` would not be fit for purpose if only one client could connect to the container port at a time. It definitely is able to concurrently proxy multiple connections. https://github.com/moby/moby/blob/27b322e56bc04da8e431fe792c742f7b52df9c46/cmd/docker-proxy/tcp_proxy.go#L77-L84

robmry replied:
Oh right, of course ... I was thinking about the one listener vs proxying multiple ports.

robmry replied:
The point being, the per-connection / traffic carrying sockets returned by Accept were already nonblocking and there's only one listener.

corhere replied:
Right, right; only one thread would be blocking on the `accept` syscall per listener. Fair point. Still, saving a thread per non-idle proxy process at the cost of one line of code with no ongoing maintenance or cognitive-complexity costs is well worth doing. And arguably it's avoiding a resource-usage regression rather than adding a optimization.

robmry replied:
Yes, it was worth sorting out - thank you for spotting it.

corhere commented on cmd/docker-proxy/main.go:
```diff
@@ -13,11 +13,27 @@ import (
+	listenSockFd
+)
+
 func main() {
```
Since we're thinking about close-on-exec already: https://github.com/golang/go/blob/1985c0ccf9ef0736aeb79ef548689aa935fa5c4c/src/syscall/exec_linux.go#L611-L612

I know that docker-proxy does not exec, and probably won't in the foreseeable future, but passing fd 3 or 4 to a child process would result in such subtle lifecycle bugs (e.g. parent pipe isn't closed until grandchild exits, listen socket remains bound until grandchild exits) that I feel it's in our best interest to preemptively fix it.

```suggestion
func main() {
	// Mark any files we expect to inherit as close-on-exec
	// so that they are not unexpectedly inherited by any child processes
	// if we ever need docker-proxy to exec something.
	// This is safe to do even if the fd belongs to the Go runtime
	// as it would be a no-op:
	// the Go runtime marks all file descriptors it opens as close-on-exec.
	// See the godoc for syscall.ForkLock for more information.
	syscall.CloseOnExec(parentPipeFd)
	syscall.CloseOnExec(listenSockFd)
```

robmry replied:
Ok

corhere commented on cmd/docker-proxy/main.go:
```diff
@@ -13,11 +13,27 @@ import (
+	}
 
-	p, err := NewProxy(host, container)
+	f := os.NewFile(parentPipeFd, "signal-parent")
```
While we're on the topic of non-blocking fds, I am tempted to set the parentPipeFd to non-blocking mode. It is a premature optimization, I admit -- we will never fill up the kernel's pipe buffer with how little we are writing today -- but it's so trivial to implement, and so obviously correct and risk-free! And if other people are going to use our code as an example to copy from, they'd end up with an implementation that is free from surprising performance characteristics.
```suggestion
	_ = runtime.SetNonblock(parentPipeFd, true)
	f := os.NewFile(parentPipeFd, "signal-parent")
```

robmry replied:
Ok

corhere reviewed (approved):
LGTM! I have a couple of suggestions regarding close-on-exec and non-blocking mode for passed fds that would be more future-proofing than anything.

This PR needs a rebase. There's a merge conflict.

AkihiroSuda commented:
Needs rebase

robmry committed:
Only build docker-proxy for Linux

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
`docker-proxy` is only used on Linux, and the latest suggested changes won't work on Windows without some Windows-specific code (the `syscall` functions are different).

So, the latest push tries to only build it for Linux - will see if CI's ok with that. Even if it is, I'm not sure whether it'll cause any packaging issues.

thaJeztah commented:
Two quick questions;


- The location of the userland-proxy binary is configurable; do we know if there's existing alternative implementations that try to match ABI compatibility? I guess we may fail if we'd encounter one of those, as they don't know about the new `use-listen-fd` option? Have we tried what this situation would look like (e.g. by using a v27.1.1 version of the proxy with this PR)?
- ☝️ I do expect this to be a really niche case, but mostly wondering how loud it fails, and if the error is clear enough for users in such a situation to trace back the cause.
- Slightly similar; how do we handle proxies when `live-restore` is enabled? Do we keep those running? And in that case, what happens when updating docker (thus: updating to a newer version of `docker-proxy`) ?
- ☝️ Note that we do _NOT_ support live-restore when doing "major version updates" (only patch releases), so depending on what version update this lands in may be just "expected behavior", but just thought I'd ask. (And to get an understanding what amount of release-notes are needed to warn users about these).

thaJeztah commented:
Also FWIW; regardless what version this PR lands in, the first commit would likely be fine on its own; it's an unused binary, and with https://github.com/docker/docker-ce-packaging/pull/1045 in place, means that our packaging scripts also no longer expect the unused binary to be present.

robmry commented:
> * The location of the userland-proxy binary is configurable; do we know if there's existing alternative implementations that try to match ABI compatibility? I guess we may fail if we'd encounter one of those, as they don't know about the new `use-listen-fd` option? Have we tried what this situation would look like (e.g. by using a v27.1.1 version of the proxy with this PR)?

An old version of our `docker-proxy` fails with a long-winded error message that ends with `failed to start docker-proxy, check that the current version is in your $PATH.` (There's a full version in the PR description.)

But no, I don't know if anyone's implemented their own version. If they have, and they ignore the new `--use-listen-fd` option, it'll probably fail because the ports are already bound... and the error response from their version of the proxy command is under their control.

> * Slightly similar; how do we handle proxies when `live-restore` is enabled? Do we keep those running? And in that case, what happens when updating docker (thus: updating to a newer version of `docker-proxy`) ?

We don't keep them running ... which seems pretty wrong. But, at least it makes this simpler.

thaJeztah reviewed (approved):
LGTM

thanks for answering my questions!
